### PR TITLE
[SPARK-50153][SQL] Add `name` to `RuleExecutor` to make printing `QueryExecutionMetrics`'s logs clearer

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to add `name` to `RuleExecutor` to make printing `QueryExecutionMetrics`'s logs clearer. 
Otherwise, the following printing is meaningless (without knowing that `RuleExecutor`'s metric is being output)
```shell
24/10/29 15:12:33 WARN PlanChangeLogger:
=== Metrics of Executed Rules ===
Total number of runs: 100
Total time: 0.8585 ms
Total number of effective runs: 0
Total time of effective runs: 0.0 ms

24/10/29 15:12:33 WARN PlanChangeLogger:
=== Metrics of Executed Rules ===
Total number of runs: 196
Total time: 0.78946 ms
Total number of effective runs: 0
Total time of effective runs: 0.0 ms
```

### Why are the changes needed?
There are many `similar outputs` printed in the log, but it seems difficult for `spark developers` to know which `RuleExecutor` generated them.

- Before:
```shell
=== Metrics of Executed Rules ===
Total number of runs: 199
Total time: 1.394873 ms
Total number of effective runs: 2
Total time of effective runs: 0.916459 ms


=== Metrics of Executed Rules ===
Total number of runs: 196
Total time: 0.525134 ms
Total number of effective runs: 0
Total time of effective runs: 0.0 ms


=== Metrics of Executed Rules ===
Total number of runs: 1
Total time: 0.00175 ms
Total number of effective runs: 0
Total time of effective runs: 0.0 ms


=== Metrics of Executed Rules ===
Total number of runs: 166
Total time: 0.876414 ms
Total number of effective runs: 1
Total time of effective runs: 0.130166 ms


=== Metrics of Executed Rules ===
Total number of runs: 1
Total time: 0.007375 ms
Total number of effective runs: 0
Total time of effective runs: 0.0 ms
```

- After:
```shell
=== Metrics of Executed Rules org.apache.spark.sql.internal.BaseSessionStateBuilder$$anon$2 ===
Total number of runs: 199
Total time: 32.982158 ms
Total number of effective runs: 2
Total time of effective runs: 32.067459 ms


=== Metrics of Executed Rules org.apache.spark.sql.internal.BaseSessionStateBuilder$$anon$2 ===
Total number of runs: 196
Total time: 0.630705 ms
Total number of effective runs: 0
Total time of effective runs: 0.0 ms


=== Metrics of Executed Rules org.apache.spark.sql.catalyst.expressions.codegen.package$ExpressionCanonicalizer ===
Total number of runs: 1
Total time: 0.105459 ms
Total number of effective runs: 0
Total time of effective runs: 0.0 ms


=== Metrics of Executed Rules org.apache.spark.sql.hive.HiveSessionStateBuilder$$anon$1 ===
Total number of runs: 166
Total time: 2.308457 ms
Total number of effective runs: 1
Total time of effective runs: 1.22025 ms


=== Metrics of Executed Rules org.apache.spark.sql.catalyst.expressions.codegen.package$ExpressionCanonicalizer ===
Total number of runs: 1
Total time: 0.009166 ms
Total number of effective runs: 0
Total time of effective runs: 0.0 ms
```

### Does this PR introduce _any_ user-facing change?
Yes, When `Spark developers` observe the logs printed by `PlanChangeLogger#logMetrics`, their meaning becomes clearer.

### How was this patch tested?
Manually check

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by panbingkun** [comment]
cc @cloud-fan @MaxGekk

**Review by MaxGekk — APPROVED** [review]
It looks helpful.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala [hunk 1]
```diff
@@ -118,6 +118,12 @@ class PlanChangeLogger[TreeType <: TreeNode[_]] extends Logging {
 
 abstract class RuleExecutor[TreeType <: TreeNode[_]] extends Logging {
 
+  /** Name for this rule executor, automatically inferred based on class name. */
+  val name: String = {
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala:122]: How about to convert it to a method with `protected` to allow override in sub-classes for better names.

**Review by HyukjinKwon — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala (hunk 1 above)
Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala:122]: Make sense, let me update it now.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @panbingkun and @HyukjinKwon for review.

**Review by dongjoon-hyun — COMMENTED** [review]
+1, late LGTM. Thank you, @panbingkun and all!

**Comment by panbingkun** [comment]
Thanks all @MaxGekk @HyukjinKwon @dongjoon-hyun ❤️ !

**Pull request closed without merging**